src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +243,190 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+                else
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
+                    return null;
+                }
+            }
+            else if (serverNamePartsByComma.Length < 2 && serverNamePartsByFowardSlash.Length == 2)
+            {
+                string hostName = serverNamePartsByFowardSlash[0];
+                string instanceName = serverNamePartsByFowardSlash[1];
+                int portNumber = -1;
+
+                if (GetPortByInstanceName(hostName, instanceName, out portNumber))
                 {
-                    portNumber = ushort.Parse(serverAndPortParts[1]);
+                    return new SNITCPHandle(hostName, portNumber, timerExpire, callbackObject, parallel);
                 }
-                catch (Exception e)
+                else
                 {
-                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InvalidConnStringError, e);
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                     return null;
                 }
             }
-            else if (serverAndPortParts.Length > 2)
+            else
             {
                 SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
                 return null;
             }
+        }
+
+        /// <summary>
+        /// Finds IPAddress by given hostname.
+        /// </summary>
+        /// <param name="hostname">hostname string</param>
+        /// <param name="hostIPAddress">host IPAddress found by hostname string</param>
+        /// <param name="family">address family. Default value is AddressFamily.InterNetwork</param>
+        /// <returns>bool value showing if operation is successful</returns>
+        private static bool GetIPAdress(string hostname, out IPAddress hostIPAddress, AddressFamily family = AddressFamily.InterNetwork)
+        {
+            hostIPAddress = null;
+
+            if (String.IsNullOrWhiteSpace(hostname))
+            {
+                return false;
+            }
+
+            try
+            {
+                Task<IPAddress[]> addressTask = Dns.GetHostAddressesAsync(hostname);
+                IPAddress[] addresses = null;
+                if (!addressTask.Wait(1000) || (addresses = addressTask.Result) == null || addresses.Length == 0)
+                {
+                    return false;
+                }
+
+                List<IPAddress> addressList = new List<IPAddress>(addresses);
+                hostIPAddress = addressList.First(a => a.AddressFamily == family);
+                if (hostIPAddress == null)
+                {
+                    return false;
+                }
+            }
+            catch
```

Comment:
We should be propagating these exceptions back to the user. There's an overload for the SNIError constructor that takes an exception. There are examples of it elsewhere in the code (like in TcpHandle).